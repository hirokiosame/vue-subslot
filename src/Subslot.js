import { emit } from './utils/emit';
import { filterVnodes } from './utils/filter-vnodes';
import createDefineMixin from './create-define-mixin';

export default {
	functional: true,
	props: {
		not: {
			type: Boolean,
		},
		element: {
			type: [Object, Array, String],
		},
		offset: {
			type: [String, Number],
			default: 0,
			validator: val => !Number.isNaN(parseInt(val, 10)),
		},
		limit: {
			type: [String, Number],
			validator: val => !Number.isNaN(parseInt(val, 10)),
		},
		vnodes: {
			type: null,
		},
		name: {
			type: String,
			default: 'default',
		},
	},

	render(h, ctx) {
		const { props, parent } = ctx;

		let vnodes;

		// Detect definition
		if (parent.$subslots) {
			vnodes = parent.$subslots[props.name];
		} else {
			vnodes = ctx.props.vnodes || ctx.parent.$slots.default || [];
			vnodes = filterVnodes({
				vnodes,
				filter: props,
				vm: parent,
			});
		}

		if (!vnodes || vnodes.length === 0) {
			emit(ctx, 'no-match');
			return ctx.slots().default;
		}

		return vnodes;
	},

	/* Static method for mixin */
	define: createDefineMixin,
};
