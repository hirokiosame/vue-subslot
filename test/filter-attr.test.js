import {mount} from '@vue/test-utils';
import Subslot from 'vue-subslot';
import CardHeader from './fixtures/CardHeader.vue';

describe('Subslot', () => {
	test('Should support filter attributes', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot element="@CardHeader" limit="1" />
					</div>

					<div class="card-content">
						<subslot not element="@CardHeader" />
					</div>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<card-header>
						Header
					</card-header>

					Content
				</card>
			`,
			components: {
				Card,
				CardHeader,
			},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support element as a direct reference', () => {
		const Card = {
			template: `
				<div class="card">
					<div class="card-header">
						<subslot :element="CardHeader" limit="1" />
					</div>

					<div class="card-content">
						<subslot not :element="CardHeader" />
					</div>
				</div>
			`,

			components: {
				Subslot,
			},

			data() {
				return {
					CardHeader,
				};
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<card-header>
						Header
					</card-header>

					Content
				</card>
			`,
			components: {
				Card,
				CardHeader,
			},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should render all', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot />
				</div>
			`,

			components: {
				Subslot,
			},

			data() {
				return {
					CardHeader,
				};
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<div>Should render</div>
					<span>Should render</span>
					<button>Should render</button>
				</card>
			`,
			components: {Card},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support tag', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot element="button" limit="1" />
				</div>
			`,

			components: {
				Subslot,
			},

			data() {
				return {
					CardHeader,
				};
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<div>Shouldn't render</div>
					<span>Shouldn't render</span>
					<button>Should render</button>
				</card>
			`,
			components: {Card},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should support wildcard', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot element="*" />
				</div>
			`,

			components: {
				Subslot,
			},

			data() {
				return {
					CardHeader,
				};
			},
		};

		const wrapper = mount({
			template: `
				<card>
					Shouldn't render
					<span>Should render</span>
					Shouldn't render
				</card>
			`,
			components: {Card},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should only be 3 CardHeaders', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot
						element="@CardHeader"
						limit="3"
					/>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<card-header
						v-for="i in 10"
						:key="i"
					>
						Header
					</card-header>
				</card>
			`,
			components: {
				Card,
				CardHeader,
			},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should enforce offset of 3', () => {
		const Card = {
			template: `
				<div class="card">
					<subslot
						element="@CardHeader"
						offset="3"
						limit="3"
					/>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<card-header
						v-for="i in 10"
						:key="i"
					>
						Header {{ i }}
					</card-header>
				</card>
			`,
			components: {
				Card,
				CardHeader,
			},
		});
		expect(wrapper.element).toMatchSnapshot();
	});

	test('Should enforce offset and limit without element', () => {
		const Card = {
			template: `
				<div>
					<subslot
						offset="3"
						limit="3"
					/>
				</div>
			`,

			components: {
				Subslot,
				CardHeader,
			},
		};

		const wrapper = mount({
			template: `
				<card>
					<div
						v-for="i in 10"
						:key="i"
					>
						Header {{ i }}
					</div>
				</card>
			`,
			components: {
				Card,
			},
		});
		expect(wrapper.element).toMatchSnapshot();
	});
});
