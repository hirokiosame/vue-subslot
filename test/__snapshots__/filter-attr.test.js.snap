// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subslot Should enforce offset and limit without element 1`] = `
<div>
  <div>
    
						Header 4
					
  </div>
  <div>
    
						Header 5
					
  </div>
  <div>
    
						Header 6
					
  </div>
</div>
`;

exports[`Subslot Should enforce offset of 3 1`] = `
<div
  class="card"
>
  <div>
    
						Header 4
					
  </div>
  <div>
    
						Header 5
					
  </div>
  <div>
    
						Header 6
					
  </div>
</div>
`;

exports[`Subslot Should only be 3 CardHeaders 1`] = `
<div
  class="card"
>
  <div>
    
						Header
					
  </div>
  <div>
    
						Header
					
  </div>
  <div>
    
						Header
					
  </div>
</div>
`;

exports[`Subslot Should render all 1`] = `
<div
  class="card"
>
  <div>
    Should render
  </div>
   
  <span>
    Should render
  </span>
   
  <button>
    Should render
  </button>
</div>
`;

exports[`Subslot Should support element as a direct reference 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    

					Content
				
  </div>
</div>
`;

exports[`Subslot Should support filter attributes 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <div>
      
						Header
					
    </div>
  </div>
   
  <div
    class="card-content"
  >
    

					Content
				
  </div>
</div>
`;

exports[`Subslot Should support tag 1`] = `
<div
  class="card"
>
  <button>
    Should render
  </button>
</div>
`;

exports[`Subslot Should support tag 2`] = `
<div class="card">
  Should render
</div>
`;

exports[`Subslot Should support wildcard 1`] = `
<div
  class="card"
>
  <span>
    Should render
  </span>
</div>
`;
